Services/Concurrency/RatedCounter.cs

```diff
@@ -0,0 +1,160 @@
+        // Note: currently, the memory usage depends on the length of the period to
+        //       monitor, so this is a good place for future optimizations
+        private readonly Queue<long> timestamps;
+
+        // Keep track of the last element enqueued, to avoid O(n) lookups
+        private long lastTimestamp;
+
+        public RatedCounter(double rate, double timeUnitLength, string name, ILogger logger)
+        {
+            if (rate < MIN_RATE)
+            {
+                var msg = "The counter rate value must be greater than or equal to " + MIN_RATE;
+                this.log.Error(msg, () => new { name, rate, timeUnitLength });
+                throw new InvalidConfigurationException(msg);
+            }
+
+            if (timeUnitLength < MIN_TIME_UNIT)
+            {
+                var msg = "The counter time unit value must be greater than or equal to " + MIN_TIME_UNIT;
+                this.log.Error(msg, () => new { name, rate, timeUnitLength });
+                throw new InvalidConfigurationException(msg);
+            }
+
+            this.name = name;
+            this.log = logger;
+
+            this.eventsPerTimeUnit = rate;
+            this.timeUnitLength = timeUnitLength;
+
+            // e.g. 60000 / 100 = 600 msecs = one event every 0.6 secs
+            this.eventInterval = timeUnitLength / rate;
+
+            this.timestamps = new Queue<long>();
+            this.lastTimestamp = 0;
+
+            this.log.Debug("New counter", () => new { name, rate, timeUnitLength });
+        }
+
+        // Increase the counter, taking a pause if the caller is going too fast.
+        // Return a boolean indicating whether a pause was required.
+        public async Task<bool> IncreaseAsync()
+        {
+            long pause = 0;
+
+            // Note: keep the code fast, e.g. leave ASAP and don't I/O while locking
+            lock (this.timestamps)
+            {
+                long now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
+
+                // Remove old events from the queue
+                var oneTimeUnitAgo = now - this.timeUnitLength;
+                while (this.timestamps.Count > 0 && this.timestamps.Peek() <= oneTimeUnitAgo)
+                {
+                    this.timestamps.Dequeue();
+                }
+
+                // How many events happened in the last time unit, and how many
+                // will happen next (the queue can contain future timestamps)
+                var count = this.timestamps.Count;
+                if (count < this.eventsPerTimeUnit)
+                {
+                    this.lastTimestamp = now;
+                    this.timestamps.Enqueue(now);
+                    return false;
+                }
+
+                // Calculate when the next event is allowed to run, i.e. how long to pause
+                var nextTime = (long) (this.lastTimestamp + this.timeUnitLength + this.eventInterval * (count - this.eventsPerTimeUnit));
```

Comment:
use now instead of this.lastTimestamp?

Reply:
`this.lastTimestamp` is the time when the last event in the queue can be executed, which is most likely in the future